### PR TITLE
Small refactor to ttnn-to-emitc pass

### PR DESCRIPTION
**Pull request by svuckovicTT**
- Refactor so it fits original llvm repo style
- Change pass base class from `func::FuncOp` to `mlir::ModuleOp`

### CODE REVIEW

File: lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                                       /*isStandard=*/false);
+    }
+
+    // TTNN -> EmitC
+    //
+    {
+      TTNNToEmitCTypeConverter typeConverter(&getContext());
+      RewritePatternSet patterns(&getContext());
+
+      // Func dialect handling
```
Comment (odjuricicTT) [inline lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp:77]: What happens if you treat the FucnOp as legal and pass it down without any conversion?

File: lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp [hunk 2]
```diff
@@ -28,6 +26,63 @@ using namespace mlir::tt;
...
 namespace {
 
+template <typename SrcOp, typename Adaptor = typename SrcOp::Adaptor>
+class DefaultOpConversionPattern : public OpConversionPattern<SrcOp> {
+  using OpConversionPattern<SrcOp>::OpConversionPattern;
+
+public:
+  // Default op conversion pattern, used to convert most ops
+  //
+  DefaultOpConversionPattern(MLIRContext *ctx)
```
Comment (odjuricicTT) [inline lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp:36]: Is this needed?

**Review by odjuricicTT — COMMENTED** [review]
Is there anything left to change from the torch style implementation? E.g. is PassDetail.h needed?

File: lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp (hunk 1 above)
Comment (svuckovicTT) [reply lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp:77]: That should work, per emitc docs, they can parse it: https://mlir.llvm.org/docs/Dialects/EmitC/

However, I think I tried earlier, and was hitting some "materialization" errors which I didn't quite understand. Let me try again.

Comment (svuckovicTT) [reply lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp:77]: With the following diff
```diff
diff --git a/lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp b/lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp
index e42e50f..b00ad00 100644
--- a/lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp
+++ b/lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp
@@ -15,6 +15,7 @@

 #include "ttmlir/Dialect/TTNN/IR/TTNN.h"
 #include "ttmlir/Dialect/TTNN/IR/TTNNOps.h"
+#include <mlir/Dialect/Func/IR/FuncOps.h>

 using namespace mlir;
 using namespace mlir::tt;
@@ -42,6 +43,8 @@ struct ConvertTTNNToEmitCPass
     target.addLegalDialect<emitc::EmitCDialect>();
     target.addIllegalDialect<ttnn::TTNNDialect>();
     target.addLegalOp<mlir::ModuleOp>();
+    target.addLegalOp<func::FuncOp>();
+    target.addLegalOp<func::ReturnOp>();

     // Add header imports to front of module
     //
@@ -74,20 +77,20 @@ struct ConvertTTNNToEmitCPass
       TTNNToEmitCTypeConverter typeConverter(&getContext());
       RewritePatternSet patterns(&getContext());

-      // Func dialect handling
-      //
-      populateFunctionOpInterfaceTypeConversionPattern<func::FuncOp>(
-          patterns, typeConverter);
-      target.addDynamicallyLegalOp<func::FuncOp>([&](func::FuncOp op) {
-        return typeConverter.isSignatureLegal(op.getFunctionType()) &&
-               typeConverter.isLegal(&op.getBody());
-      });
-      populateReturnOpTypeConversionPattern(patterns, typeConverter);
-      target.addDynamicallyLegalOp<func::ReturnOp>(
-          [&](func::ReturnOp op) { return typeConverter.isLegal(op); });
-      populateCallOpTypeConversionPattern(patterns, typeConverter);
-      target.addDynamicallyLegalOp<func::CallOp>(
-          [&](func::CallOp op) { return typeConverter.isLegal(op); });
+      // // Func dialect handling
+      // //
+      // populateFunctionOpInterfaceTypeConversionPattern<func::FuncOp>(
+      //     patterns, typeConverter);
+      // target.addDynamicallyLegalOp<func::FuncOp>([&](func::FuncOp op) {
+      //   return typeConverter.isSignatureLegal(op.getFunctionType()) &&
+      //          typeConverter.isLegal(&op.getBody());
+      // });
+      // populateReturnOpTypeConversionPattern(patterns, typeConverter);
+      // target.addDynamicallyLegalOp<func::ReturnOp>(
+      //     [&](func::ReturnOp op) { return typeConverter.isLegal(op); });
+      // populateCallOpTypeConversionPattern(patterns, typeConverter);
+      // target.addDynamicallyLegalOp<func::CallOp>(
+      //     [&](func::CallOp op) { return typeConverter.isLegal(op); });

       // TTNN -> EmitC patterns
       //
```

if I run
```bash
./build/bin/ttmlir-opt --ttir-layout --ttnn-open-device --convert-ttir-to-ttnn --convert-ttnn-to-emitc test/ttmlir/Dialect/TTNN/simple_multiply.mlir | ./build/bin/ttmlir-translate -mlir-to-cpp -allow-unregistered-dialect
```

I get

```cpp
test/ttmlir/Dialect/TTNN/simple_multiply.mlir:10:10: error: failed to legalize unresolved materialization from 'tensor<64x128xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<64x128xf32, #tt.memory_space<system>>>>' to '!emitc.opaque<"ttnn::Tensor">' that remained live after conversion
    %1 = "ttir.multiply"(%arg0, %arg1, %0) <{operandSegmentSizes = array<i32: 2, 1>, operand_constraints = [#any_device, #any_device, #any_device]}> : (tensor<64x128xf32>, tensor<64x128xf32>, tensor<64x128xf32>) -> tensor<64x128xf32>
         ^
test/ttmlir/Dialect/TTNN/simple_multiply.mlir:10:10: note: see current operation: %0 = "builtin.unrealized_conversion_cast"(%arg1) : (tensor<64x128xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<64x128xf32, #tt.memory_space<system>>>>) -> !emitc.opaque<"ttnn::Tensor">
test/ttmlir/Dialect/TTNN/simple_multiply.mlir:10:10: note: see existing live user here: %10 = emitc.call_opaque "ttnn::to_memory_config"(%0, %8) : (!emitc.opaque<"ttnn::Tensor">, !emitc.opaque<"ttnn::Tensor">) -> !emitc.opaque<"ttnn::Tensor">
```

Comment (svuckovicTT) [reply lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp:77]: I think I understand what's going on... If I don't have actual conversion patterns added to `RewritePatternSet`, operands of func ops won't be converted. Having these calls

```cpp
      populateFunctionOpInterfaceTypeConversionPattern<func::FuncOp>(
          patterns, typeConverter);
      populateReturnOpTypeConversionPattern(patterns, typeConverter);
      populateCallOpTypeConversionPattern(patterns, typeConverter);
```

takes care of that.

File: lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp (hunk 2 above)
Comment (svuckovicTT) [reply lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp]: No, removed.

File: lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp (hunk 1 above)
Comment (rpavlovicTT) [reply lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp:78]: Reminder that we can delete "PassDetail.h" header and move `GEN_PASS_DEF_CONVERTTTNNTOEMITC` step into the .cpp file.

File: lib/Conversion/PassDetail.h
Comment (nsmithtt) [inline lib/Conversion/PassDetail.h:1]: Seems like we can remove this file?

File: lib/Conversion/TTNNToEmitC/PopulatePatterns.h
Comment (nsmithtt) [inline lib/Conversion/TTNNToEmitC/PopulatePatterns.h:1]: Same with this one?

File: lib/Conversion/TTNNToEmitC/TypeConverter.h
Comment (nsmithtt) [inline lib/Conversion/TTNNToEmitC/TypeConverter.h:1]: Same with this one

**Review by nsmithtt — CHANGES_REQUESTED** [review]

File: lib/Conversion/PassDetail.h
Comment (nsmithtt) [reply lib/Conversion/PassDetail.h:1]: Nvm, the way GitHub displays this diff it looks like all of the lines were deleted, but the file still remains. I now see on the LHS the icon showing it's removed

**Review by nsmithtt — APPROVED** [review]

**Review by rpavlovicTT — APPROVED** [review]
